google-cloud-bigtable/lib/google/cloud/bigtable/column_family_map.rb

```diff
@@ -0,0 +1,341 @@
+      #   column_families.add "cf2", gc_rule: Google::Cloud::Bigtable::GcRule.max_versions(5)
+      #
+      class ColumnFamilyMap
+        include Enumerable
+
+        ##
+        # Create a new ColumnFamilyMap object.
+        #
+        # @example
+        #   require "google/cloud/bigtable"
+        #
+        #   bigtable = Google::Cloud::Bigtable.new
+        #
+        #   column_families = Google::Cloud::Bigtable::ColumnFamilyMap.new
+        #
+        #   column_families.add "cf1", gc_rule: Google::Cloud::Bigtable::GcRule.max_age(600)
+        #   column_families.add "cf2", gc_rule: Google::Cloud::Bigtable::GcRule.max_versions(5)
+        #
+        # @example Create new families using block syntax:
+        #   require "google/cloud/bigtable"
+        #
+        #   bigtable = Google::Cloud::Bigtable.new
+        #
+        #   column_families = Google::Cloud::Bigtable::ColumnFamilyMap.new do |cfm|
+        #     cfm.add "cf1", gc_rule: Google::Cloud::Bigtable::GcRule.max_age(600)
+        #     cfm.add "cf2", gc_rule: Google::Cloud::Bigtable::GcRule.max_versions(5)
+        #   end
+        #
+        def initialize
+          @column_families = empty_map
+
+          yield self if block_given?
+        end
+
+        def each
+          return enum_for :each unless block_given?
+
+          @column_families.each do |name, column_family|
+            yield name, ColumnFamily.from_grpc(column_family, name)
+          end
+        end
+
+        def [] name
+          return nil unless name? name
+
+          ColumnFamily.from_grpc(@column_families[name], name)
+        end
+
+        def name? name
+          @column_families.has_key? name
+        end
+
+        def names
+          @column_families.keys
+        end
+
+        def length
+          @column_families.length
+        end
+        alias size length
+
+        def empty?
+          length.zero?
+        end
+
+        ##
+        # Adds a new column family to the table.
+        #
+        # @param name [String] Column family name
+        # @param gc_rule [Google::Cloud::Bigtable::GcRule] The garbage
+        #   collection rule to be used for the column family. Optional. The
+        #   service default value will be used when not specified.
+        #
+        # @raise [ArgumentError] if the column family name already exists.
+        # @raise [FrozenError] if the column family map is frozen.
+        #
+        # @example
+        #   require "google/cloud/bigtable"
+        #
+        #   bigtable = Google::Cloud::Bigtable.new
+        #
+        #   table = bigtable.table("my-instance", "my-table", perform_lookup: true)
+        #
+        #   table.column_families do |column_families|
+        #     column_families.add "cf4", gc_rule: Google::Cloud::Bigtable::GcRule.max_age(600)
+        #     column_families.add "cf5", gc_rule: Google::Cloud::Bigtable::GcRule.max_versions(5)
+        #   end
+        #
+        def add name, gc_rule: nil
```

Comment:
If this was a new feature, I would be in favor of `gc_rule` as a named argument, since it may be `nil`, and other attributes may be added later to `ColumnFamily`. However, making this change here breaks backward compatibility for the `#create_table` block.

Reply:
Yes. But when we discussed this earlier I remember you being on board with this change. Since we are already breaking backwards compatibility, this seems like a good time to make this change. I changed this because it is consistent with my recollection of our previous discussions about it.

Reply:
I think it would be fine for the block to `Table#column_families` since that is new, but I don't think it's worth making a breaking change to the existing `#create_table` block. Breaking changes are a judgement call, that just my personal opinion.

Reply:
I would still make the change. If you want to maintain backwards compatibility you could get fancy and change the method signature to `def add name, positional_gc_rule = nil, gc_rule: nil` and then use `positional_gc_rule` when it had a value, and print a deprecation warning. That would allow the old calls to still work while pointing users to the new API.

I would use `@overload` for the arguments in the documentation to only include `name` and `gc_rule` tho.

Reply:
Updated this PR with these changes.